### ISSUE TITLE
Garbage collection output does not list the pack IDs

### ISSUE THREAD

**Issue by mattleibow** [dotnet/sdk#20350]
After the installation, the GC runs and outputs all the things it removes. Except, it just puts out version numbers:

> note the double space that seems to indicate it should have been there...

```
Garbage collecting for SDK feature band(s) 6.0.100...
Uninstalling workload pack  version 6.0.100-rc.1.1516.
Uninstalling workload pack  version 15.0.100-rc.1.1528.
Uninstalling workload pack  version 15.0.100-rc.1.1528.
Uninstalling workload pack  version 15.0.100-rc.1.1528.
Uninstalling workload pack  version 15.0.100-rc.1.1528.
Uninstalling workload pack  version 15.0.100-rc.1.1528.
Uninstalling workload pack  version 12.0.100-rc.1.1528.
Uninstalling workload pack  version 12.0.100-rc.1.1528.
Uninstalling workload pack  version 12.0.100-rc.1.1528.
Uninstalling workload pack  version 12.0.100-rc.1.1528.
Uninstalling workload pack  version 12.0.100-rc.1.1528.
Uninstalling workload pack  version 6.0.100-rc.1.1516.
Uninstalling workload pack  version 6.0.100-rc.1.1516.
```

**Comment by KalleOlaviNiemitalo** [comment]
It should come from PackInfo.Id:

<https://github.com/dotnet/sdk/blob/1f6482d1ed33083dadea53c0336091fb65652513/src/Cli/dotnet/commands/dotnet-workload/install/LocalizableStrings.resx#L264-L266>
<https://github.com/dotnet/sdk/blob/1f6482d1ed33083dadea53c0336091fb65652513/src/Cli/dotnet/commands/dotnet-workload/install/NetSdkManagedInstaller.cs#L414>
<https://github.com/dotnet/sdk/blob/24654cc1cae4f52414ba095926f6f2f63a2a739e/src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/WorkloadResolver.cs#L585-L588>

[#18166](https://github.com/dotnet/sdk/pull/18166/commits/3d64056448b16a822d7826b8b279769858bf6b0a#diff-3ed786bee415b3fd9a38f33a2d47f86a1d8bfc2e7a58f866b9d73b80d57f6604L515-R513) changed the type of PackInfo.Id from `string` to [WorkloadPackId](https://github.com/dotnet/sdk/blob/1f6482d1ed33083dadea53c0336091fb65652513/src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/WorkloadPackId.cs). Might that have affected the JSON format so that installation records saved by earlier versions can no longer be fully deserialized?

<https://github.com/dotnet/sdk/blob/1f6482d1ed33083dadea53c0336091fb65652513/src/Cli/dotnet/commands/dotnet-workload/install/NetSdkManagedInstaller.cs#L330-L336>

**Comment by KalleOlaviNiemitalo** [comment]
I ran `dotnet workload install wasm-tools` on .NET SDK 6.0.100-preview.7.21379.14 and got a `C:\Program Files\dotnet\metadata\workloads\InstalledPacks\v1\Microsoft.NET.Runtime.Emscripten.Node\6.0.0-preview.7.21377.2\6.0.100` file containing:

```JSON
{"Id":{},"Version":"6.0.0-preview.7.21377.2","Kind":0,"ResolvedPackageId":"Microsoft.NET.Runtime.Emscripten.2.0.23.Node.win-x64","Path":"C:\\Program Files\\dotnet\\packs\\Microsoft.NET.Runtime.Emscripten.2.0.23.Node.win-x64\\6.0.0-preview.7.21377.2","IsStillPacked":true}
```

i.e. the ID string was lost already in JSON serialization and not just in deserialization. I presume this is because [WorkloadPackId](https://github.com/dotnet/sdk/blob/1f6482d1ed33083dadea53c0336091fb65652513/src/Resolvers/Microsoft.NET.Sdk.WorkloadManifestReader/WorkloadPackId.cs) does not have any public properties. So if you want to use WorkloadPackId in PackInfo, then it needs JSON converters or alternative string-typed properties.

**Comment by sfoslund** [comment]
I'm not able to repro this issue with out latest build. I tried out installing/ garbage collecting android and wasm-tools packs and both showed up with the correct pack ids. @KalleOlaviNiemitalo @mattleibow can you give more details about your configuration? It looks like you're on windows using preview 7 file based installs, right? Are you seeing it with all workloads or only specific ones?

**Comment by KalleOlaviNiemitalo** [comment]
@sfoslund, I suspected that @mattleibow was hitting a JSON deserialization problem, installed `wasm-tools` to examine the workload metadata file format, and found that the `"Id"` property was clearly incomplete, which could explain the problem. This was with .NET SDK installed from dotnet-sdk-6.0.100-preview.7.21379.14-win-x64.exe. I have not tried other workloads. I don't know how to trigger the garbage collection itself and have not verified that the pack ID is missing from the output there.

When you install a workload, do you get this empty `"Id":{}` or something else?

**Comment by sfoslund** [comment]
Ah, you're right, the id is empty but it wasn't repro-ing as it was going down a different code path. I have enough to investigate now, thanks!

**Comment by sfoslund** [comment]
Fixed by #20876
